src/PooledArrays.jl

```diff
@@ -341,16 +465,22 @@ Base.@propagate_inbounds function Base.isassigned(pa::PooledArray, I::Int...)
     !iszero(pa.refs[I...])
 end
 
-# Vector case
-Base.@propagate_inbounds function Base.getindex(A::PooledArray, I::Union{Real,AbstractVector}...)
-    PooledArray(RefArray(getindex(A.refs, I...)), copy(A.invpool))
+# Other cases; we rely on the fact that non-standard indexing will fall back to this eventually
+Base.@propagate_inbounds function Base.getindex(A::PooledArray, I::Union{Real, AbstractVector}...)
+    # make sure we do not increase A.refcount in case creation of newrefs fails
+    newrefs = getindex(A.refs, I...)
+    @assert newrefs isa AbstractArray
```

Comment:
Why the `@assert` here? Are we worried this might ever be something non-AbstractArray?

Reply:
@bkamins likes checking that unexpected things don't happen. ;-) In this case the compile will get rid of it anyway so I guess it's free.

Reply:
Yes - @nalimilan knows my style well. It should be optiimzed out, and otherwise we throw early and otherwise we should add a comment to line `PooledArray(RefArray(newrefs), A.invpool, A.pool, A.refcount)` with an obscure error that `newrefs` is expected to be an `AbstractArray`. And the `@assert` is self documenting.